### PR TITLE
Update for Ember Data 1.13.x, closes #32

### LINKED ISSUE #32: Update for Ember 2.0

### PR DESCRIPTION
**Pull request by hhff**

### CODE REVIEW

**Comment by hhff** [comment]
@truenorth / @kellyselden can you have a peeper at this?

**Comment by kellyselden** [comment]
I'll try and find some today.

File: addon/mixins/route.js [hunk 1]
```diff
@@ -115,7 +125,15 @@ export default Ember.Mixin.create({
   */
   infinityModel(modelName, options) {
 
-    if (Ember.isEmpty(this.store) || Ember.isEmpty(this.store.find)){
+    if (emberDataVersionIs('greaterThan', '1.0.0-beta.19.2') && emberDataVersionIs('lessThan', '1.13.4')) {
```
Comment (kellyselden) [inline addon/mixins/route.js]: Why are you excluding 1.13.0 through 1.13.3?

**Comment by kellyselden** [comment]
You could test the find/query resolution. I would do it by putting `_emberDataVersionIs: emberDataVersionIs` on the route, change references in the route, then inject a new version in a test
```js
_emberDataVersionIs: function(operation, version) {
  if (operation === 'lessThan' && version === '1.13.0') {
    return true;
  }
}
```
then assert it's calling `store.find` instead of `store.query`. You could also test your error handling this way.

File: addon/mixins/route.js (hunk 1 above)
Comment (hhff) [inline addon/mixins/route.js]: There was a regression with how it was unpacking `meta` from the payload in those versions, causing Ember Infinity to think all of the records were loaded immediately.

**Comment by hhff** [comment]
Thanks @kellyselden - will add some test coverage for that

File: addon/mixins/route.js (hunk 1 above)
Comment (truenorth) [inline addon/mixins/route.js]: We should outline all of the ember-infinity use cases and ensure that ember-data tests are in place, to protect against future incompatibilities. Skipping many versions of a library can make upgrading apps difficult.

Comment (hhff) [inline addon/mixins/route.js]: I agree.  I wouldn't normally do something like this^^  but the excluded versions of Ember Data were all released within a period of 3 days.  Figured anyone on one of those versions could trivially update or downgrade easily, given they're so close (and given they're genuinely bugged, that developer probably should update too).

**Comment by hhff** [comment]
I've been holding off finishing this PR because I've had some requests about how this lib could be used without ember data.  Instead of dynamically setting the store find method, I've been playing around with the idea of allowing the user to actually set the 'promise generator' themselves.  

Ember Data probably shouldn't be a dependency!  I think that's where we could could start the 0.1.x track

**Comment by kellyselden** [comment]
@hhff good ideas.

**Comment by kellyselden** [comment]
@hhff but if the goal of this PR is to ease deprecation warning suffering, maybe it should not be included in the extracting ember data plans?

**Comment by hhff** [comment]
watchu mean @kellyselden ?

**Comment by kellyselden** [comment]
The goal of this PR was to git rid of the deprecations from `store.find`, not taking out ember-data?

**Comment by hhff** [comment]
yup!  just figured the logic was somewhat related.  I agree though - I'll get this in first

File: addon/mixins/route.js (hunk 1 above)
Comment (mike-north) [inline addon/mixins/route.js]: :

**Comment by hhff** [comment]
@kellyselden @mike-north do u guys know why this type of thing happens with Ember Try?  Of all of things that have made trying to support 1.x & 2.x tracks of Ember difficult, this Ember Try module loading is by far the hardest to deal with.

https://travis-ci.org/hhff/ember-infinity/jobs/76094856#L446

**Comment by mike-north** [comment]
@hhff I'll take a look. I'm very familiar with ember-try :)

**Comment by hhff** [comment]
TY @mike-north !  Figured we'd have to do something like this - you're the man.  Will squash now and lets get this bad boy innnnn

**Comment by hhff** [comment]
@mike-north / @kellyselden can one of you guys give this a final once over and drop merge if you're happy

**Comment by kellyselden** [comment]
@hhff I don't have too much time to deep dive, but it looks good at a glance.

File: addon/mixins/route.js [hunk 2]
```diff
@@ -1,7 +1,7 @@
 import Ember from 'ember';
+import { emberDataVersionIs } from 'ember-version-is';
 
-const { get } = Ember;
-
+const keys = Object.keys ? Object.keys : Ember.keys;
```
Comment (mike-north) [inline addon/mixins/route.js]: Consider
```js
const keys = Object.keys || Ember.keys;
```

File: addon/mixins/route.js [hunk 3]
```diff
@@ -245,8 +262,8 @@ export default Ember.Mixin.create({
    @return {Object}
    */
   _includeBoundParams: function(options, boundParams) {
-    if (Ember.keys(boundParams).length > 0) {
-      Ember.keys(boundParams).forEach( (key) => {
+    if (keys(boundParams).length > 0) {
```
Comment (mike-north) [inline addon/mixins/route.js]: I would stick to using 
```js
Ember.isEmpty()
```
as you have done here https://github.com/hhff/ember-infinity/pull/48/files#diff-95eb5c8003a4fd239be76137499e5176R146

File: addon/mixins/route.js [hunk 4]
```diff
@@ -245,8 +262,8 @@ export default Ember.Mixin.create({
...
+      keys(boundParams).forEach( (key) => {
```
Comment (mike-north) [inline addon/mixins/route.js]: Just a style issue, but parens around `(key)` are not strictly needed. You could even one-line it if you wanted to

```js
keys(boundParams).forEach(k => options[k] = this.get(boundParams[k])
```

File: config/ember-try.js [hunk 5]
```diff
@@ -7,46 +7,53 @@ module.exports = {
...
+      name: 'ember-1.13',
+      dependencies: {
+        ember: '~1.13.0'
       }
     },
     {
       name: 'ember-release',
       dependencies: {
-        'ember': 'components/ember#release'
+        ember: 'components/ember#release'
```
Comment (mike-north) [inline config/ember-try.js]: I'd add an `'ember-data': '~2.0.0'` dependency to everything from release through canary

File: tests/dummy/app/adapters/application.js [hunk 6]
```diff
@@ -0,0 +1,5 @@
+import DS from 'ember-data';
+
+export default DS.RESTAdapter.extend({
+  shouldBackgroundReloadRecord: function() { return false; }
```
Comment (mike-north) [inline tests/dummy/app/adapters/application.js]: the cool kids are using ES6 function syntax now :)

```js
shouldBackgroundReloadRecord() { return false; }
```

File: tests/unit/mixins/route-test.js [hunk 7]
```diff
@@ -440,3 +440,83 @@ test('it allows overrides/manual invocations of updateInfinityModel', assert =>
...
+  DS.VERSION = "1.0.0-beta.19.3";
+  
+  var infinityError;
+  try {
+    route.model();
+  } catch(error) {
+    infinityError = error;
+  }
+  
+  assert.equal(infinityError.message, "Ember Infinity: You are using an unsupported version of Ember Data.  Please upgrade to at least 1.13.4 or downgrade to 1.0.0-beta.19.2");
```
Comment (mike-north) [inline tests/unit/mixins/route-test.js]: Consider using `QUnit.throws()` ( https://api.qunitjs.com/throws/ ) for this test

```js
QUnit.throws(() => {
  route.model();
}, /unsupported version of Ember Data/);
```

**Comment by mike-north** [comment]
@hhff review complete

**Comment by hhff** [comment]
on it @mike-north thankyou!

File: tests/dummy/app/adapters/application.js (hunk 6 above)
Comment (hhff) [inline tests/dummy/app/adapters/application.js]: i'm cool!!!!

**Comment by hhff** [comment]
@mike-north - actioned all the things, can you press the big green button?

**Comment by hhff** [comment]
w00t

**Pull request merged**